### PR TITLE
0x05C native and java method tracing

### PR DESCRIPTION
**Pull request by su-vikas**
for issue #1360 and #1361 

@cpholguera I have not incorporated all your suggestions in order to keep it concise. Let me know if is still okay or still need them?

### CODE REVIEW

**Comment by cpholguera** [comment]
FYI: I've started a review already and will finish it during this week :)

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 1]
```diff
@@ -953,6 +953,62 @@ Jprobes and Kretprobes are other KProbes-based probe types that allow hooking of
...
+In theory, Frida can be used to perform method tracing for Java code. Currently, Frida [does not](https://github.com/frida/frida-python/issues/70 "does not") have any out of the box tool for Java method tracing, unlike native code tracing for which `frida-trace` is available (which we will discuss shortly in next section). In order to perform Java method tracing, it requires manual scripting, which is nothing else but hooking the function.
+
+##### Native Code Tracing
+
+Native methods tracing can be performed with relative ease than compared to Java method tracing. `frida-trace` is a CLI tool for dynamically tracing function calls. It makes tracing native functions trivial and can be very useful for collecting information about an application.
+
+In order to use `frida-trace`, a frida server should be running on the device. An example for tracing libc's `open` function using `frida-trace` is demonstrated below, where `i` option specifies the function to be traced.
+
+```bash
+$ frida-trace -i "open" -U com.android.chrome
```
Comment (cpholguera) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: I'd put the -U first as in the other examples, to remark that's independent from com.andr...
```suggestion
$ frida-trace -U -i "open" com.android.chrome
```

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 2]
```diff
@@ -953,6 +953,62 @@ Jprobes and Kretprobes are other KProbes-based probe types that allow hooking of
...
+$ frida-trace -U -i "Java_*" com.android.chrome
+```
+
+Many binaries are stripped and don't have function name symbols available with them. In such cases, a function can be traced using its address as well.
+
+```bash
+$ frida-trace -p 1372 -a "libjpeg.so!0x4793c"
+```
+
+`frida-trace` is a very versatile tool and there are multiple configuration options available for advance usage. To learn more about them, check the [documentation](https://frida.re/docs/frida-trace/ "documentation").
```
Comment (cpholguera) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: I moved the first sentence above.
```suggestion
To learn more about all options for advanced usage, check the [documentation in the official Frida website](https://frida.re/docs/frida-trace/ "documentation").
```

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 3]
```diff
@@ -953,6 +953,62 @@ Jprobes and Kretprobes are other KProbes-based probe types that allow hooking of
...
+      log('\t return: ' + retval);      \\ edited
+  }
+}
+```
+
+In the above script, `onLeave` event can be edited to print the return values.
+
+Another thing to notice in the output above is the colorised output. An application can have multiple threads running, and each thread can call `open` function independently. By using such a color scheme, the output can be easily visually segregated for each thread.
+
+`frida-trace` can also be used to trace JNI functions in an Android application:
```
Comment (cpholguera) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: From here on I'd put a little list replacing 998-1008:
```suggestion
`frida-trace` is a very versatile tool and there are multiple configuration options available such as:
- Including `-I` and excluding `-X` entire modules.
- Tracing all JNI functions in an Android application using `-i "Java_*"` (note the use of a glob `*` to match all possible functions starting with "Java_").
- Tracing functions by address when no function name symbols are available (stripped binaries), e.g. `-a "libjpeg.so!0x4793c"`.
```

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 4]
```diff
@@ -953,6 +953,62 @@ Jprobes and Kretprobes are other KProbes-based probe types that allow hooking of
...
+  },
+
+
+  onLeave: function (log, retval, state) {
+      log('\t return: ' + retval);      \\ edited
+  }
+}
+```
+
+In the above script, `onLeave` event can be edited to print the return values.
```
Comment (cpholguera) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]: I've enhanced this line with onEnter and a note for when frida-trace cannot get the params itself.
```suggestion
In the above script, `onEnter` takes care of logging the calls to this function and its two input parameters in the right format. You can edit the `onLeave` event to print the return values as shown above.

> Note that libc is a well-known library, Frida is able to derive the input parameters of its `open` function and automatically log them correctly. But this won't be the case for other libraries or for Android Kotlin/Java code. In that case, you may want to obtain the signatures of the functions you're interested in by referring to Android Developers documentation or by reverse engineer the app first.
```

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 5]
```diff
@@ -953,6 +953,62 @@ Jprobes and Kretprobes are other KProbes-based probe types that allow hooking of
...
+
+Another thing to notice in the output above is the colorised output. An application can have multiple threads running, and each thread can call `open` function independently. By using such a color scheme, the output can be easily visually segregated for each thread.
```
Comment (cpholguera) [inline Document/0x05c-Reverse-Engineering-and-Tampering.md]:
```suggestion
Another thing to notice in the output above is the colorized output. An application can have multiple threads running, and each thread can call `open` function independently. By using such a color scheme, the output can be easily visually segregated for each thread.
```

File: Document/0x05c-Reverse-Engineering-and-Tampering.md [hunk 6]
```diff
@@ -953,6 +953,62 @@ Jprobes and Kretprobes are other KProbes-based probe types that allow hooking of
 
 The stock Android kernel comes without loadable module support, which is a problem because Kprobes are usually deployed as kernel modules. The strict memory protection the Android kernel is compiled with is another issue because it prevents the patching of some parts of Kernel memory. Elfmaster's system call hooking method causes a Kernel panic on stock Lollipop and Marshmallow because the sys_call_table is non-writable. You can, however, use KProbes in a sandbox by compiling your own, more lenient Kernel (more on this later).
 
+##### Java Method Tracing
+
+Method profiling, introduced in the previous sections, can be used to obtain the sequence in which methods are being called. This can provide us hints about critical functions of an application. But profiling has it owns limitation, it does not provide us the information about the input arguments and return values of these critical functions. For determining the input and output values of a function, along with how frequently it is being called, technique of function tracing can be very useful. Tracing can also prove to be very useful while dealing with obfuscated application or applications with big codebase.
```
[the remark this answers is not in the document]
Comment (su-vikas) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: Yep, this is good. Few corrections:
- helps you "to" also determine 
- and/or "are" obfuscated -> being doesn't sound right to me, you agree?

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 5 above)
Comment (su-vikas) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: My bad, I didn't follow the style guide.

**Comment by su-vikas** [comment]
@cpholguera thanks for the suggestions, i have accepted them and suggested minor fixes for one of them. Thanks.

File: Document/0x05c-Reverse-Engineering-and-Tampering.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (cpholguera) [reply Document/0x05c-Reverse-Engineering-and-Tampering.md]: I agree on both and have changed the suggestion, thank you!

**Pull request merged**